js/main.js

```diff
@@ -689,12 +679,13 @@ function initiate() {
 
         /* STAGE: Text in math mode */
         for (var i = 0; i < mathFragments.length; ++i) {
-            var badPos = mathFragments[i].search(new RegExp('[' + smallCyrLetters + capCyrLetters + ']'));
-            if (badPos < 0){
-                badPos = mathFragments[i].search(/([^a-z\\]|^)[a-z]{4,}/i);
-            }
-            if (badPos >= 0 && mathFragments[i].search(/\\(text|mbox|hbox)/) < 0 && mathFragments[i].substr(0, badPos).search(/\\label/) < 0) {
-                addTypicalWarning('TEXT_IN_MATH_MODE', 'math', i, badPos);
+            var errorRegex = /(?:[^a-z\\{]|^)[a-zАБВГДЕЁЖЗИКЛМНОПРСТУФХЦЧШЩЬЫЪЭЮЯабвгдеёжзиклмнопрстуфхцчшщьыъэюя]{4,}/ig;
+            var result = errorRegex.exec(mathFragments[i]);
+            while (result !== null) {
+                if (!/\\(text|mbox|hbox|label)\s*\{(?!.*?\})/.test(mathFragments[i].substring(0, result.index))) {
```

Comment:
Yes, sorry, one more time I used lookahead for no reason. Changed.
As for the first question: `$\text     {hey}$` is totally valid. So is `$$\text \n {hi}$$`.